### PR TITLE
[CD-173] Auto-close nav when a menu loses keyboard focus

### PR DESCRIPTION
**Pull request by rupl**
## Types of changes

- :heavy_check_mark: Improvement (non-breaking change which iterates on an existing feature)
- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)
- Breaking change (fix or feature that would cause existing functionality to not work as expected)
- Security update (dependency updates, or to fix a vulnerability)

## Description
All of our dropdowns can stay open (at least until another is opened) and that sometimes made keyboard nav feel a little clunky. It sometimes hides content that has focus, and made screen reading a little strange as well.

The new behavior closes the dropdown once focus has completely left both the dropdown and its related toggler.  I had to attach all the listeners to the document to include the toggler, but overall it was the best, most reliable behavior with no edge cases that I could produce as I smashed the keyboard

This applies to any dropdown: OCHA Services, Language, User menu, Search, Main nav

## Steps to reproduce the problem or Steps to test

  1. Load branch, `drush cr`
  2. Use keyboard to tab forwards and backwards through page, opening menus and observing their behavior as focus changes.

## Impact
No impact. I also moved the `cd-user-menu.js` file to its own component in this branch, but grepping the codebase indicates we never directly reference it. All references were twig-style includes:

```
{{ attach_library('common_design/cd-user-menu') }}
```

## PR Checklist

- [x] I have followed the Conventional Commits guidelines.

### CODE REVIEW

**Review by orakili — CHANGES_REQUESTED** [review]
This works well for toggleable menus but messes up non menu toggleable elements like the [show full description](https://reliefweb.int/disaster/eq-2021-000116-hti) on reliefweb.int or the [show all donors](https://www.unocha.org/our-funding) on unocha.org.

This is because, after clicking the button, the focus moves the title of the toggleable section which is outside of the section (so that it's always visible regardless of the state of the toggleable section).

So I think we need an attribute for example to disable this behavior.

Alternatively, one ugly solution that is compatible with those sites is to wrap the focusout event handler(s) in a `if (!element.hasAttribute('data-cd-insert-after')) {`.

This `data-cd-insert-after` attribute is already loaded as discussed in https://github.com/UN-OCHA/common_design/pull/420 and this `focusout` is not unrelated to the notion of `has-popup` so maybe that would be ok.

File: components/cd-dropdown/cd-dropdown.js [hunk 1]
```diff
@@ -360,6 +358,16 @@
       toggler.addEventListener('keydown', this.handleEscape);
       element.addEventListener('keydown', this.handleEscape);
 
+      // When keyboard focus leaves this menu, close it automatically.
+      //
+      // We listen to the document so that backward tabbing can still close the
+      // menu, but it will stay open if focus lands on the related toggler.
+      document.addEventListener('focusout', (ev) => {
```
[the remark this answers is not in the document]
Comment (rupl) [reply components/cd-dropdown/cd-dropdown.js:365]: Great solution! I'll update the branch.

**Comment by rupl** [comment]
As discussed in-person during retreat, we're accepting the regression for the disclosure pattern temporarily, then adding a new CD Disclosure component for those other use-cases.

File: components/cd-dropdown/cd-dropdown.js [hunk 2]
```diff
@@ -356,10 +354,26 @@
...
+      });
+
+      // Hide the menu when the focus moves from the toggler to
+      // an element outside the menu.
+      toggler.addEventListener('focusout', (ev) => {
+        if (ev.relatedTarget && !element.contains(ev.relatedTarget) && element !== ev.relatedTarget) {
+          this.toggle(toggler, true);
+        }
+      });
+
```
Comment (orakili) [inline components/cd-dropdown/cd-dropdown.js:376]:
```suggestion
      // Hide the dropdown menu when focus changes.
      // Do not apply when the menu is not a dropdown which is currently the
      // case, in practice, when using the 'data-cd-insert-after' attribute.
      // @todo remove when we introduce the Disclosure component.
      if (!element.hasAttribute('data-cd-insert-after')) {
        // Hide the menu when the focus moves from inside the menu to
        // an element outside the menu or its toggler.
        element.addEventListener('focusout', (ev) => {
          if (ev.relatedTarget && !element.contains(ev.relatedTarget) && toggler !== ev.relatedTarget) {
            this.toggle(toggler, true);
          }
        });

        // Hide the menu when the focus moves from the toggler to
        // an element outside the menu.
        toggler.addEventListener('focusout', (ev) => {
          if (ev.relatedTarget && !element.contains(ev.relatedTarget) && element !== ev.relatedTarget) {
            this.toggle(toggler, true);
          }
        });
      }
```

@rupl The suggestion just wraps the event listeners in a condition that ensures that they are not set for `disclosure` cases so that there is no breaking change, leaving us more time to create the disclosure component.
